File: twidere/src/main/java/org/mariotaku/twidere/provider/CacheProvider.java
package org.mariotaku.twidere.provider;

import android.content.ContentProvider;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.ParcelFileDescriptor;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringDef;
import android.webkit.MimeTypeMap;

import com.bluelinelabs.logansquare.JsonMapper;
import com.nostra13.universalimageloader.cache.disc.DiskCache;

import org.mariotaku.commons.logansquare.LoganSquareMapperFinder;
import org.mariotaku.restfu.RestFuUtils;
import org.mariotaku.twidere.TwidereConstants;
import org.mariotaku.twidere.model.CacheMetadata;
import org.mariotaku.twidere.task.SaveFileTask;
import org.mariotaku.twidere.util.BitmapUtils;
import org.mariotaku.twidere.util.dagger.GeneralComponentHelper;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Locale;

import javax.inject.Inject;

import okio.ByteString;

/**
 * Created by mariotaku on 16/1/1.
 */
public class CacheProvider extends ContentProvider implements TwidereConstants {
    @Inject
    DiskCache mSimpleDiskCache;

    public static Uri getCacheUri(String key, @Type String type) {
        final Uri.Builder builder = new Uri.Builder();
        builder.scheme(ContentResolver.SCHEME_CONTENT);
        builder.authority(TwidereConstants.AUTHORITY_TWIDERE_CACHE);
        builder.appendPath(ByteString.encodeUtf8(key).base64Url());
        if (type != null) {
            builder.appendQueryParameter(QUERY_PARAM_TYPE, type);
        }
        return builder.build();
    }

    @NonNull
    public static String getCacheKey(Uri uri) {
        if (!ContentResolver.SCHEME_CONTENT.equals(uri.getScheme()))
            throw new IllegalArgumentException(uri.toString());
        if (!TwidereConstants.AUTHORITY_TWIDERE_CACHE.equals(uri.getAuthority()))
            throw new IllegalArgumentException(uri.toString());
        return ByteString.decodeBase64(uri.getLastPathSegment()).utf8();
    }


    public static String getMetadataKey(Uri uri) {
        if (!ContentResolver.SCHEME_CONTENT.equals(uri.getScheme()))
            throw new IllegalArgumentException(uri.toString());
        if (!TwidereConstants.AUTHORITY_TWIDERE_CACHE.equals(uri.getAuthority()))
            throw new IllegalArgumentException(uri.toString());
        return getExtraKey(ByteString.decodeBase64(uri.getLastPathSegment()).utf8());
    }

    public static String getExtraKey(String key) {
        return key + ".extra";
    }

    @Override
    public boolean onCreate() {
        final Context context = getContext();
        assert context != null;
        GeneralComponentHelper.build(context).inject(this);
        return true;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {
        return null;
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        final CacheMetadata metadata = getMetadata(uri);
        if (metadata != null) {
            return metadata.getContentType();
        }
        final String type = uri.getQueryParameter(QUERY_PARAM_TYPE);
        if (type != null) {
            switch (type) {
                case Type.IMAGE: {
                    final File file = mSimpleDiskCache.get(getCacheKey(uri));
                    if (file == null) return null;
                    return BitmapUtils.getImageMimeType(file);
                }
                case Type.VIDEO: {
                    return "video/mp4";
                }
                case Type.JSON: {
                    return "application/json";
                }
            }
        }
        return null;
    }

    public CacheMetadata getMetadata(@NonNull Uri uri) {
        final File file = mSimpleDiskCache.get(getMetadataKey(uri));
        if (file == null) return null;
        FileInputStream is = null;
        try {
            final JsonMapper<CacheMetadata> mapper = LoganSquareMapperFinder.mapperFor(CacheMetadata.class);
            is = new FileInputStream(file);
            return mapper.parse(is);
        } catch (IOException e) {
            return null;
        } finally {
            RestFuUtils.closeSilently(is);
        }
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, ContentValues values) {
        return null;
    }

    @Override
    public int delete(@NonNull Uri uri, String selection, String[] selectionArgs) {
        return 0;
    }

    @Override
    public int update(@NonNull Uri uri, ContentValues values, String selection, String[] selectionArgs) {
        return 0;
    }

    @Nullable
    @Override
    public ParcelFileDescriptor openFile(@NonNull Uri uri, @NonNull String mode) throws FileNotFoundException {
        try {
            final File file = mSimpleDiskCache.get(getCacheKey(uri));
            if (file == null) throw new FileNotFoundException();
            final int modeBits = modeToMode(mode);
            if (modeBits != ParcelFileDescriptor.MODE_READ_ONLY)
                throw new IllegalArgumentException("Cache can't be opened for write");
            return ParcelFileDescriptor.open(file, modeBits);
        } catch (IOException e) {
            throw new FileNotFoundException();
        }
    }

    /**
     * Copied from ContentResolver.java
     */
    private static int modeToMode(String mode) {
        int modeBits;
        if ("r".equals(mode)) {
            modeBits = ParcelFileDescriptor.MODE_READ_ONLY;
        } else if ("w".equals(mode) || "wt".equals(mode)) {
            modeBits = ParcelFileDescriptor.MODE_WRITE_ONLY
                    | ParcelFileDescriptor.MODE_CREATE
                    | ParcelFileDescriptor.MODE_TRUNCATE;
        } else if ("wa".equals(mode)) {
            modeBits = ParcelFileDescriptor.MODE_WRITE_ONLY
                    | ParcelFileDescriptor.MODE_CREATE
                    | ParcelFileDescriptor.MODE_APPEND;
        } else if ("rw".equals(mode)) {
            modeBits = ParcelFileDescriptor.MODE_READ_WRITE
                    | ParcelFileDescriptor.MODE_CREATE;
        } else if ("rwt".equals(mode)) {
            modeBits = ParcelFileDescriptor.MODE_READ_WRITE
                    | ParcelFileDescriptor.MODE_CREATE
                    | ParcelFileDescriptor.MODE_TRUNCATE;
        } else {
            throw new IllegalArgumentException("Invalid mode: " + mode);
        }
        return modeBits;
    }

    public static final class CacheFileTypeCallback implements SaveFileTask.FileInfoCallback {
        private final Context context;
        private final String type;

        public CacheFileTypeCallback(Context context, @Type String type) {
            this.context = context;
            this.type = type;
        }

        @Override
        @NonNull
        public String getFilename(@NonNull Uri source) {
            String cacheKey = getCacheKey(source);
            final int indexOfSsp = cacheKey.indexOf("://");
            if (indexOfSsp != -1) {
                cacheKey = cacheKey.substring(indexOfSsp + 3);
            }
            return cacheKey.replaceAll("[^\\w\\d_]", String.valueOf(getSpecialCharacter()));
        }

        @Override
        @Nullable
        public String getMimeType(@NonNull Uri source) {
            if (type == null || source.getQueryParameter(QUERY_PARAM_TYPE) != null) {
                return context.getContentResolver().getType(source);
            }
            final Uri.Builder builder = source.buildUpon();
            builder.appendQueryParameter(QUERY_PARAM_TYPE, type);
            return context.getContentResolver().getType(builder.build());
        }

        @Override
        public String getExtension(@Nullable String mimeType) {
            if (mimeType == null) return null;
            return MimeTypeMap.getSingleton().getExtensionFromMimeType(mimeType.toLowerCase(Locale.US));
        }

        @Override
        public char getSpecialCharacter() {
            return '_';
        }
    }

    @StringDef({Type.IMAGE, Type.VIDEO, Type.JSON})
    public @interface Type {
        String IMAGE = "image";
        String VIDEO = "video";
        String JSON = "json";
    }
}

File: twidere/src/main/kotlin/org/mariotaku/twidere/provider/CacheProvider.kt
package org.mariotaku.twidere.provider

import android.content.ContentProvider
import android.content.ContentResolver
import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.ParcelFileDescriptor
import okio.ByteString.Companion.decodeBase64
import okio.ByteString.Companion.encodeUtf8
import org.mariotaku.mediaviewer.library.FileCache
import org.mariotaku.twidere.TwidereConstants.AUTHORITY_TWIDERE_CACHE
import org.mariotaku.twidere.TwidereConstants.QUERY_PARAM_TYPE
import org.mariotaku.twidere.annotation.CacheFileType
import org.mariotaku.twidere.model.CacheMetadata
import org.mariotaku.twidere.task.SaveFileTask
import org.mariotaku.twidere.util.JsonSerializer
import org.mariotaku.twidere.util.dagger.GeneralComponent
import java.io.ByteArrayInputStream
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream
import javax.inject.Inject

/**
 * Created by mariotaku on 16/1/1.
 */
class CacheProvider : ContentProvider() {
    @Inject
    internal lateinit var fileCache: FileCache

    override fun onCreate(): Boolean {
        GeneralComponent.get(context!!).inject(this)
        return true
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?,
            selectionArgs: Array<String>?, sortOrder: String?): Cursor? {
        return null
    }

    override fun getType(uri: Uri): String? {
        val metadata = getMetadata(uri)
        if (metadata != null) {
            return metadata.contentType
        }
        when (uri.getQueryParameter(QUERY_PARAM_TYPE)) {
            CacheFileType.IMAGE -> {
                val file = fileCache.get(getCacheKey(uri)) ?: return null
                return BitmapFactory.Options().apply {
                    inJustDecodeBounds = true
                    BitmapFactory.decodeFile(file.absolutePath, this)
                }.outMimeType
            }
            CacheFileType.VIDEO -> {
                return "video/mp4"
            }
            CacheFileType.JSON -> {
                return "application/json"
            }
        }
        return null
    }

    fun getMetadata(uri: Uri): CacheMetadata? {
        val bytes = fileCache.getExtra(getCacheKey(uri)) ?: return null
        return try {
            ByteArrayInputStream(bytes).use {
                return@use JsonSerializer.parse(it, CacheMetadata::class.java)
            }
        } catch (e: IOException) {
            null
        }
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        return null
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        return 0
    }

    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int {
        return 0
    }

    @Throws(FileNotFoundException::class)
    override fun openFile(uri: Uri, mode: String): ParcelFileDescriptor? {
        try {
            val file = fileCache.get(getCacheKey(uri)) ?: throw FileNotFoundException()
            val modeBits = modeToMode(mode)
            if (modeBits != ParcelFileDescriptor.MODE_READ_ONLY)
                throw IllegalArgumentException("Cache can't be opened for write")
            return ParcelFileDescriptor.open(file, modeBits)
        } catch (e: IOException) {
            throw FileNotFoundException()
        }

    }

    class ContentUriFileInfo(
            private val context: Context,
            private val uri: Uri,
            @CacheFileType override val cacheFileType: String?
    ) : SaveFileTask.FileInfo, CacheFileTypeSupport {
        override val fileName: String by lazy {
            var cacheKey = getCacheKey(uri)
            val indexOfSsp = cacheKey.indexOf("://")
            if (indexOfSsp != -1) {
                cacheKey = cacheKey.substring(indexOfSsp + 3)
            }
            return@lazy cacheKey.replace("[^\\w\\d_]".toRegex(), specialCharacter.toString())
        }

        override val mimeType: String? by lazy {
            if (cacheFileType == null || uri.getQueryParameter(QUERY_PARAM_TYPE) != null) {
                return@lazy context.contentResolver.getType(uri)
            }
            val builder = uri.buildUpon()
            builder.appendQueryParameter(QUERY_PARAM_TYPE, cacheFileType)
            return@lazy context.contentResolver.getType(builder.build())
        }

        override val specialCharacter: Char
            get() = '_'

        override fun inputStream(): InputStream {
            return context.contentResolver.openInputStream(uri)!!
        }

        override fun close() {
            // No-op
        }
    }

    interface CacheFileTypeSupport {
        val cacheFileType: String?
    }


    companion object {

        fun getCacheUri(key: String, @CacheFileType type: String?): Uri {
            val builder = Uri.Builder()
            builder.scheme(ContentResolver.SCHEME_CONTENT)
            builder.authority(AUTHORITY_TWIDERE_CACHE)
            builder.appendPath(key.encodeUtf8().base64Url())
            if (type != null) {
                builder.appendQueryParameter(QUERY_PARAM_TYPE, type)
            }
            return builder.build()
        }

        fun getCacheKey(uri: Uri): String {
            if (ContentResolver.SCHEME_CONTENT != uri.scheme)
                throw IllegalArgumentException(uri.toString())
            if (AUTHORITY_TWIDERE_CACHE != uri.authority)
                throw IllegalArgumentException(uri.toString())
            return uri.lastPathSegment?.decodeBase64()!!.utf8()
        }


        /**
         * Copied from ContentResolver.java
         */
        private fun modeToMode(mode: String): Int {
            return if ("r" == mode) {
                ParcelFileDescriptor.MODE_READ_ONLY
            } else if ("w" == mode || "wt" == mode) {
                ParcelFileDescriptor.MODE_WRITE_ONLY or ParcelFileDescriptor.MODE_CREATE or ParcelFileDescriptor.MODE_TRUNCATE
            } else if ("wa" == mode) {
                ParcelFileDescriptor.MODE_WRITE_ONLY or ParcelFileDescriptor.MODE_CREATE or ParcelFileDescriptor.MODE_APPEND
            } else if ("rw" == mode) {
                ParcelFileDescriptor.MODE_READ_WRITE or ParcelFileDescriptor.MODE_CREATE
            } else if ("rwt" == mode) {
                ParcelFileDescriptor.MODE_READ_WRITE or ParcelFileDescriptor.MODE_CREATE or ParcelFileDescriptor.MODE_TRUNCATE
            } else {
                throw IllegalArgumentException("Invalid mode: $mode")
            }
        }
    }
}
